netty/src/main/java/io/grpc/netty/ProtocolNegotiators.java

```diff
@@ -167,6 +175,72 @@ public FromChannelCredentialsResult withCallCredentials(CallCredentials callCred
     }
   }
 
+  public static FromServerCredentialsResult from(ServerCredentials creds) {
+    if (creds instanceof TlsServerCredentials) {
+      TlsServerCredentials tlsCreds = (TlsServerCredentials) creds;
+      Set<TlsServerCredentials.Feature> incomprehensible =
+          tlsCreds.incomprehensible(understoodServerTlsFeatures);
+      if (!incomprehensible.isEmpty()) {
+        return FromServerCredentialsResult.error(
+            "TLS features not understood: " + incomprehensible);
+      }
+      SslContextBuilder builder = GrpcSslContexts.forServer(
+          new ByteArrayInputStream(tlsCreds.getCertificateChain()),
+          new ByteArrayInputStream(tlsCreds.getPrivateKey()),
+          tlsCreds.getPrivateKeyPassword());
+      SslContext sslContext;
+      try {
+        sslContext = builder.build();
+      } catch (SSLException ex) {
+        throw new RuntimeException(
```

Comment:
Instead of throwing an exception here, why not return a result with error? This exception could be because of bad input arguments?

Reply:
I wasn't wild about this as well. The problem is that all the useful information is in the Exception and its causal chain. There's no good way to combine multiple Exceptions from different sources together without suppressed exceptions.

I was fine with throwing on bad input arguments or if the platform was messed up in some way. The main case that I think is poorly served is if the data is in a format that _this_ server can't process but some other server can. Since only the Netty server exists, that's a theoretical discussion. I figured we'd fix it up if/when it became a problem.

Thinking more about the platform issues that might throw here, I think they would be pretty rare. So it would probably be best to make this an IllegalArgumentException.

Reply:
Swapped to IllegalArgumentException.